### PR TITLE
Issue #1231 Separating the code to cache the ISO

### LINKED ISSUE #1231: Extra progress bar for downloading ISO during minishift start
If I start minishift from fresh i.e. without cached ISO, I can see below extra progress bars. 
However it is not consistently reproducible. I could reproduce it twice but after that I could not. 

Opening the issue to findout if anyone else sees it . 

Also when I looked in to the code I saw that we have spinner which includes the downloading of the ISO i.e https://github.com/minishift/minishift/blob/master/cmd/minishift/cmd/start.go#L334-L347
Ideally we should running the spinner when the VM is starting. 

Minishift version: compiled from master head i.e. a6074fb6b1a153a7e43727d05d9792d26fa69461
```
$ rm -rf ~/.minishift

$ minishift start
-- Installing default add-ons ... OK
-- Checking if KVM driver is installed ... 
   Driver is available at /usr/local/bin/docker-machine-driver-kvm ... OK
-- Starting local OpenShift cluster using 'kvm' hypervisor ...
-- Minishift VM will be configured with ...
   Memory:    2 GB
   vCPUs :    2
   Disk size: 20 GB
-- Starting Minishift VM ... 
   Downloading ISO 'https://github.com/minishift/minishift-b2d-iso/releases/download/v1.0.2/minishift-b2d.iso'
951.57 KiB / 40.00 MiB [===>----------------------------------------------------------------------------------------------------------------------------------------------------------------------]   2.32% 10m14s 1.39 MiB / 40.00 MiB [======>----------------------------------------------------------------------------------------------------------------------------------------------------------------------]   3.49% 7m45s 1.53 MiB / 40.00 MiB [======>----------------------------------------------------------------------------------------------------------------------------------------------------------------------]   3.82% 7m13s 2.27 MiB / 40.00 MiB [=========>-------------------------------------------------------------------------------------------------------------------------------------------------------------------]   5.68% 5m32s 3.04 MiB / 40.00 MiB [=============>---------------------------------------------------------------------------------------------------------------------------------------------------------------]   7.59% 4m35s 4.37 MiB / 40.00 MiB [==================>----------------------------------------------------------------------------------------------------------------------------------------------------------]  10.91% 3m47s 4.65 MiB / 40.00 MiB [====================>--------------------------------------------------------------------------------------------------------------------------------------------------------]  11.62% 3m40s 4.71 MiB / 40.00 MiB [====================>--------------------------------------------------------------------------------------------------------------------------------------------------------]  11.79% 3m40s 5.08 MiB / 40.00 MiB [=====================>-------------------------------------------------------------------------------------------------------------------------------------------------------]  12.70% 3m33s 5.23 MiB / 40.00 MiB [======================>------------------------------------------------------------------------------------------------------------------------------------------------------]  13.07% 3m30s 5.46 MiB / 40.00 MiB [=======================>-----------------------------------------------------------------------------------------------------------------------------------------------------]  13.65% 3m26s 7.01 MiB / 40.00 MiB [==============================>-----------------------------------------------------------------------------------------------------------------------------------------------]  17.51% 3m2s 7.17 MiB / 40.00 MiB [===============================>---------------------------------------------------------------------------------------------------------------------------------------------]  17.93% 2m59s 7.72 MiB / 40.00 MiB [=================================>-------------------------------------------------------------------------------------------------------------------------------------------]  19.30% 2m54s 8.32 MiB / 40.00 MiB [===================================>-----------------------------------------------------------------------------------------------------------------------------------------]  20.79% 2m53s 9.88 MiB / 40.00 MiB [==========================================>----------------------------------------------------------------------------------------------------------------------------------]  24.69% 2m38s 10.46 MiB / 40.00 MiB [============================================>-------------------------------------------------------------------------------------------------------------------------------]  26.15% 2m33s 10.72 MiB / 40.00 MiB [==============================================>-----------------------------------------------------------------------------------------------------------------------------]  26.81% 2m31s 10.79 MiB / 40.00 MiB [==============================================>-----------------------------------------------------------------------------------------------------------------------------]  26.98% 2m31s 11.12 MiB / 40.00 MiB [===============================================>----------------------------------------------------------------------------------------------------------------------------]  27.81% 2m29s 11.40 MiB / 40.00 MiB [=================================================>--------------------------------------------------------------------------------------------------------------------------]  28.51% 2m27s 11.90 MiB / 40.00 MiB [===================================================>------------------------------------------------------------------------------------------------------------------------]  29.76% 2m23s 12.04 MiB / 40.00 MiB [===================================================>------------------------------------------------------------------------------------------------------------------------]  30.09% 2m22s 12.33 MiB / 40.00 MiB [=====================================================>----------------------------------------------------------------------------------------------------------------------]  30.84% 2m20s 12.40 MiB / 40.00 MiB [=====================================================>----------------------------------------------------------------------------------------------------------------------]  31.00% 2m19s 13.43 MiB / 40.00 MiB [=========================================================>------------------------------------------------------------------------------------------------------------------]  33.58% 2m11s 14.06 MiB / 40.00 MiB [============================================================>----------------------------------------------------------------------------------------------------------------]  35.15% 2m8s 14.58 MiB / 40.00 MiB [===============================================================>-------------------------------------------------------------------------------------------------------------]  36.44% 2m5s 15.29 MiB / 40.00 MiB [==================================================================>----------------------------------------------------------------------------------------------------------]  38.22% 2m1s 15.56 MiB / 40.00 MiB [==================================================================>---------------------------------------------------------------------------------------------------------]  38.89% 1m59s 17.45 MiB / 40.00 MiB [===========================================================================>------------------------------------------------------------------------------------------------]  43.62% 1m55s 18.19 MiB / 40.00 MiB [==============================================================================>---------------------------------------------------------------------------------------------]  45.49% 1m56s 18.24 MiB / 40.00 MiB [==============================================================================>---------------------------------------------------------------------------------------------]  45.61% 1m56s 18.38 MiB / 40.00 MiB [===============================================================================>--------------------------------------------------------------------------------------------]  45.94% 1m55s 19.29 MiB / 40.00 MiB [==================================================================================>-----------------------------------------------------------------------------------------]  48.23% 1m51s 20.12 MiB / 40.00 MiB [======================================================================================>-------------------------------------------------------------------------------------]  50.30% 1m46s 20.49 MiB / 40.00 MiB [========================================================================================>-----------------------------------------------------------------------------------]  51.21% 1m44s 21.35 MiB / 40.00 MiB [===========================================================================================>--------------------------------------------------------------------------------]  53.37% 1m38s 21.43 MiB / 40.00 MiB [============================================================================================>-------------------------------------------------------------------------------]  53.58% 1m38s 29.60 MiB / 40.00 MiB [================================================================================================================================>---------------------------------------------]  74.00% 52s 30.10 MiB / 40.00 MiB [==================================================================================================================================>-------------------------------------------]  75.25% 49s 30.31 MiB / 40.00 MiB [===================================================================================================================================>------------------------------------------]  75.79% 48s 30.85 MiB / 40.00 MiB [======================================================================================================================================>---------------------------------------]  77.11% 45s 31.08 MiB / 40.00 MiB [=======================================================================================================================================>--------------------------------------]  77.69% 44s 32.04 MiB / 40.00 MiB [===========================================================================================================================================>----------------------------------]  80.10% 39s 33.00 MiB / 40.00 MiB [===============================================================================================================================================>------------------------------]  82.51% 34s 33.67 MiB / 40.00 MiB [==================================================================================================================================================>---------------------------]  84.17% 30s 36.11 MiB / 40.00 MiB [=============================================================================================================================================================>----------------]  90.27% 18s 38.78 MiB / 40.00 MiB [=========================================================================================================================================================================>-----]  96.95% 5s 39.06 MiB / 40.00 MiB [==========================================================================================================================================================================>----]  97.66% 4s 40.00 MiB / 40.00 MiB [===============================================================================================================================================================================] 100.00%- 
OK
-- Checking for IP address ... OK
-- Checking if external host is reachable from the Minishift VM ... 
   Pinging 8.8.8.8 ... OK
-- Checking HTTP connectivity from the VM ... 
   Retrieving http://minishift.io/index.html ... OK
-- Checking if persistent storage volume is mounted ... OK
-- Checking available disk space ... 0% OK
-- Downloading OpenShift binary 'oc' version 'v3.6.0'
 34.72 MiB / 34.72 MiB [===============================================================================================================================================================================] 100.00% 0s
-- OpenShift cluster will be configured with ...
   Version: v3.6.0
-- Checking `oc` support for startup flags ... 
   host-config-dir ... OK
   host-data-dir ... OK
   host-pv-dir ... OK
   host-volumes-dir ... OK
   routing-suffix ... OK
Starting OpenShift using openshift/origin:v3.6.0 ...
Pulling image openshift/origin:v3.6.0
Pulled 1/4 layers, 26% complete
Pulled 1/4 layers, 33% complete
Pulled 1/4 layers, 37% complete
Pulled 1/4 layers, 46% complete
```

### PR DESCRIPTION
**Pull request by LalatenduMohanty**
Fixes #1231   

Signed-off-by: Lalatendu Mohanty <lmohanty@redhat.com>

### CODE REVIEW

**Comment by gbraad** [comment]
Since we check now if host is able to run, this makes more sense.

**Comment by LalatenduMohanty** [comment]
Here is how the output looks now
```
$ minishift start
-- Checking if KVM driver is installed ... 
   Driver is available at /usr/local/bin/docker-machine-driver-kvm ... OK
-- Starting local OpenShift cluster using 'kvm' hypervisor ...
-- Minishift VM will be configured with ...
   Memory:    2 GB
   vCPUs :    2
   Disk size: 20 GB

   Downloading ISO 'https://github.com/minishift/minishift-b2d-iso/releases/download/v1.0.2/minishift-b2d.iso'
 40.00 MiB / 40.00 MiB [===============================================================================================================================================================================] 100.00% 0s
-- Starting Minishift VM ... OK
-- Checking for IP address ... OK
-- Checking if external host is reachable from the Minishift VM ... 
   Pinging 8.8.8.8 ... OK
-- Checking HTTP connectivity from the VM ... 
   Retrieving http://minishift.io/index.html ... OK
-- Checking if persistent storage volume is mounted ... OK
-- Checking available disk space ... 0% OK
-- Downloading OpenShift binary 'oc' version 'v3.6.0'
```

**Comment by praveenkumar** [comment]
Below is what I am now getting as part of this fix. 

```
-- Checking if KVM driver is installed ... 
   Driver is available at /usr/sbin/docker-machine-driver-kvm ... OK
-- Starting local OpenShift cluster using 'kvm' hypervisor ...
-- Minishift VM will be configured with ...
   Memory:    2 GB
   vCPUs :    2
   Disk size: 20 GB

   Downloading ISO 'https://github.com/minishift/minishift-centos-iso/releases/download/v1.1.0/minishift-centos7.iso'
 333.00 MiB / 333.00 MiB [=======================================================================================================================================] 100.00% 0s
-- Starting Minishift VM ... \ 
   Setting hostname ..| OK  ==> not sure why this unexpected behavior 
OK
-- Checking for IP address ... OK
-- Checking if external host is reachable from the Minishift VM ... 
   Pinging 8.8.8.8 ... OK
-- Checking HTTP connectivity from the VM ... 
   Retrieving http://minishift.io/index.html ... OK
-- Checking if persistent storage volume is mounted ... OK
-- Checking available disk space ... 1% OK
````

File: cmd/minishift/cmd/start.go [hunk 1]
```diff
@@ -330,6 +330,14 @@ func startHost(libMachineClient *libmachine.Client) *host.Host {
 		// Should handle cached images
 		//fmt.Println("   Boot ISO:  ", machineConfig.MinikubeISO)
 	}
+
+	//Cache the required ISO for Minishift
+	if machineConfig.ShouldCacheMinikubeISO() {
```
Comment (hferentschik) [inline cmd/minishift/cmd/start.go]: I thought we said we move it into a function?

File: pkg/minikube/cluster/cluster.go [hunk 2]
```diff
@@ -359,11 +359,6 @@ func (m *MachineConfig) IsMinikubeISOCached() bool {
 }
 
 func createHost(api libmachine.API, config MachineConfig) (*host.Host, error) {
-	if config.ShouldCacheMinikubeISO() {
```
Comment (hferentschik) [inline pkg/minikube/cluster/cluster.go]: No need to remove it from here. This is not our code, probably better to leave it untouched.

File: cmd/minishift/cmd/start.go [hunk 3]
```diff
@@ -330,6 +330,9 @@ func startHost(libMachineClient *libmachine.Client) *host.Host {
 		// Should handle cached images
 		//fmt.Println("   Boot ISO:  ", machineConfig.MinikubeISO)
 	}
+
+	cacheMinishiftISO(*machineConfig)
```
Comment (praveenkumar) [inline cmd/minishift/cmd/start.go]: why not `cacheMinishiftISO(*machineConfig)` => cacheMinishiftISO(machineConfig)

File: cmd/minishift/cmd/start.go [hunk 4]
```diff
@@ -635,3 +638,11 @@ func IsOpenShiftRunning(driver drivers.Driver) bool {
 
 	return openshift.IsRunning(dockerCommander)
 }
+
+func cacheMinishiftISO(config cluster.MachineConfig) {
```
Comment (praveenkumar) [inline cmd/minishift/cmd/start.go]: Here you can have `config *cluster.MachineConfig`

File: cmd/minishift/cmd/start.go (hunk 3 above)
Comment (hferentschik) [inline cmd/minishift/cmd/start.go]: it is quite uncommon to de-reference this way, also you effectively create a copy now

**Comment by hferentschik** [comment]
Rebased and merged. Thanks.

**Pull request closed without merging**